_drafts/faster-builds.md

```diff
@@ -0,0 +1,605 @@
+re-architect your project.
+
+1. Create a header file `precompiled.h` and place it in your source tree.
+2. Add a rule to compile this header, just as you would with a `.c` file. It
+   should be built _before_ building the `.c` or `.cpp` files. With Make, this
+   looks something like:
+
+   ```make
+   PRECOMPILED_H := "precompiled.h"
+   PRECOMPILED_H_OBJ := $(PRECOMPILED_H).gch
+
+   %.h.gch: $(PRECOMPILED_H)
+   	$(CC) $(CFLAGS) -c -o $@ $<
+   ```
+
+3. Include this file using a CFLAG argument `-include <>` when compiling the
+   `.c` files.
+
+   ```make
+   %.o : %.c | $(PRECOMPILED_H_OBJ)
+   	$(CC) $(CFLAGS) -include $(PRECOMPILED_H) -c -o $@ $<
+   ```
+
+Time to try it out. Let's go back to our `extra.c` file we had defined above.
+
+```c
+// extra.c
+
+#include "stm32f4xx.h"
+
+char *extra(void) {
+  return "extra";
+}
+```
+
+Recall that the inclusion of the `stm32f4xx.h` header caused the build time of
+this file to slow down by \~150 ms.
+
+If we create a `precompiled.h` file with the contents below
+
+```c
+// precompiled.h
+
+#include "stm32f4xx.h"
+
+// Add other large headers
+```
+
+and compile our `extra.c` file with this included, we go back to our 50 ms
+compile time!
+
+## Profiling the Build System
+
+I've now provided a few ways for you to speed up the compilation of files, but
+how do you go about profiling the rest of the build system rules? I have some
+good news and bad news.
+
+The good news is that most modern build systems have a way to show the execution
+trace of the entire build, broken down by each rule. This is a great way to find
+out how long each file takes to build and also to find where the bottle necks
+are during parallel builds.
+
+The bad news is that if you are using Make, there isn't a great way to do
+profiling. My ultimate suggestion would be to just not use Make and opt for
+something more modern. Bazel and CMake are good options. However, continue
+reading if you want a couple of hacks for profiling build steps with Make.
+
+### Profiling Individual Make Build Rules
+
+The easiest way to add timing information to compilation steps is to prepend the
+`time` command to the desired rules. For instance, if we wanted to profile each
+call to `$(CC)`, we could do something like
+
+```make
+CC = time arm-none-eabi-gcc
+```
+
+or
+
+```make
+%.o : %.c
+	time $(CC) $(CFLAGS) -c -o $@ $<
+
+```
+
+> NOTE: When using parallel builds in Make, the logging output can get
+> scrambled. My advice is to ensure you are using a newer version of GNU Make
+> and the `--output-sync=recurse` command line option[^13].
+
+### Profiling Every Make Build Rule
+
+If you want to profile _every_ build rule in your Makefile, a
+[clever approach](http://alangrow.com/blog/profiling-every-command-in-a-makefile)
+is to override the `SHELL` Make variable with a timing script.
+
+It produces a lot of noise, but it may help you discover slow steps in the
+build.
+
```

Comment:
fwiw, If you are trying to figure out why so much is triggering for an incremental rebuild, I think the output from `make --debug=basic` can also be helpful